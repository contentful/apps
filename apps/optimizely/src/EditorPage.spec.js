import React from 'react';
import { render, wait, configure } from '@testing-library/react';

import mockProps from './mockProps';
import mockVariantData from './mockData/mockVariantData';

import EditorPage from '../src/EditorPage';

configure({ testIdAttribute: 'data-test-id' });

describe('EditorPage', () => {
  let sdk;
  beforeEach(() => {
    sdk = mockProps.sdk;
    sdk.entry.fields.experimentId.onValueChanged = (fn) => () => {};

    Date.now = jest.fn(() => 100);
    Math.random = jest.fn(() => 0.5);
  });

  it('should show the reauth modal when no client is available', () => {
    const { getByTestId } = render(<EditorPage sdk={sdk} />);

    expect(getByTestId('reconnect-optimizely')).toMatchSnapshot();
  });

  it('should show the preemtive reconnect warning box', () => {
    const expires = (Date.now() + 50000).toString();
    const { getByTestId } = render(<EditorPage sdk={sdk} client={() => {}} expires={expires} />);

    expect(getByTestId('preemptive-connect')).toMatchSnapshot();
  });

  it('should show the experiment data when loaded', async () => {
    const space = {
      getContentTypes: () => Promise.resolve(mockVariantData.contentTypes),
      getEntries: () => Promise.resolve(mockVariantData.entries),
    };

    let valueChange = null;

    sdk = { ...sdk, space };

    sdk.entry.fields.experimentTitle.setValue = jest.fn();
    sdk.entry.fields.experimentId.onValueChanged = (fn) => {
      valueChange = fn;
      return () => {};
    };

    const client = {
      getExperiments: () => Promise.resolve(mockVariantData.experiments),
      getResultsUrl: (campaignUrl, experimentId) => {
        return `https://app.optimizely.com/v2/projects/123/results/${campaignUrl}/experiments/${experimentId}`;
      },
      getExperimentResults: () => Promise.resolve(mockVariantData.results),
    };

    const expires = (Date.now + 10000000).toString();
    const { container } = render(
      <EditorPage sdk={sdk} expires={expires} client={client} openAuth={() => {}} />
    );

    await wait();
    valueChange('15049730511');

    await wait();

    expect(container).toMatchSnapshot();
  });
});
