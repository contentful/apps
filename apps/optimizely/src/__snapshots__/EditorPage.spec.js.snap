// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorPage should show the experiment data when loaded 1`] = `
<div>
  <div
    class="css-175t7kh"
    data-test-id="editor-page"
  >
    <div
      class="css-3p38p2"
    >
      <div
        class="css-xieh1a"
      >
        <svg
          class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--positive___1V4nP css-uoder"
          data-test-id="cf-ui-icon"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <span>
          Select experiment
        </span>
      </div>
      <svg
        class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD css-13af3fo"
        data-test-id="cf-ui-icon"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M8.62623 18.7071C8.2357 18.3166 8.2357 17.6834 8.62623 17.2929L13.9191 12L8.62623 6.70711C8.2357 6.31658 8.2357 5.68342 8.62623 5.29289C9.01675 4.90237 9.64992 4.90237 10.0404 5.29289L16.0404 11.2929C16.431 11.6834 16.431 12.3166 16.0404 12.7071L10.0404 18.7071C9.64992 19.0976 9.01675 19.0976 8.62623 18.7071Z"
        />
      </svg>
      <div
        class="css-xieh1a"
      >
        <svg
          class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD css-uoder"
          data-test-id="cf-ui-icon"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
          />
        </svg>
        <span>
          Add content
        </span>
      </div>
      <svg
        class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD css-13af3fo"
        data-test-id="cf-ui-icon"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M8.62623 18.7071C8.2357 18.3166 8.2357 17.6834 8.62623 17.2929L13.9191 12L8.62623 6.70711C8.2357 6.31658 8.2357 5.68342 8.62623 5.29289C9.01675 4.90237 9.64992 4.90237 10.0404 5.29289L16.0404 11.2929C16.431 11.6834 16.431 12.3166 16.0404 12.7071L10.0404 18.7071C9.64992 19.0976 9.01675 19.0976 8.62623 18.7071Z"
        />
      </svg>
      <div
        class="css-xieh1a"
      >
        <svg
          class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD css-uoder"
          data-test-id="cf-ui-icon"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
          />
        </svg>
        <span>
          Publish variations
        </span>
      </div>
      <svg
        class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD css-13af3fo"
        data-test-id="cf-ui-icon"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M8.62623 18.7071C8.2357 18.3166 8.2357 17.6834 8.62623 17.2929L13.9191 12L8.62623 6.70711C8.2357 6.31658 8.2357 5.68342 8.62623 5.29289C9.01675 4.90237 9.64992 4.90237 10.0404 5.29289L16.0404 11.2929C16.431 11.6834 16.431 12.3166 16.0404 12.7071L10.0404 18.7071C9.64992 19.0976 9.01675 19.0976 8.62623 18.7071Z"
        />
      </svg>
      <div
        class="css-xieh1a"
      >
        <svg
          class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--positive___1V4nP css-uoder"
          data-test-id="cf-ui-icon"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <span>
          Start experiment
        </span>
      </div>
    </div>
    <hr
      class="css-14380qe"
    />
    <h2
      class="Heading__Heading___83D3K css-1fobf8d"
      data-test-id="cf-ui-heading"
    >
      Referenced in:
    </h2>
    <div
      class="css-1y5g446"
    >
      <p
        class="Paragraph__Paragraph___2aO-9 css-acwcvw"
        data-test-id="cf-ui-paragraph"
      >
        No other entries link to this entry.
      </p>
    </div>
    <hr
      class="css-14380qe"
    />
    <h2
      class="Heading__Heading___83D3K css-1fobf8d"
      data-test-id="cf-ui-heading"
    >
      Experiment:
    </h2>
    <div
      class="SelectField__SelectField___kbQlf"
      data-test-id="cf-ui-select-field"
    >
      <div
        class="SelectField__SelectField__label-wrapper___3jGwo"
      >
        <label
          class="FormLabel__FormLabel___3d6zQ"
          data-test-id="cf-ui-form-label"
          for="experiment"
        >
          Optimizely experiment
          <span
            class="FormLabel__FormLabel__required-text___3mvdm"
          >
            (
            required
            )
          </span>
        </label>
      </div>
      <div
        class="Select__Select__wrapper___2mbYV Select__Select--large___1HaJb"
      >
        <select
          aria-label="experiment"
          class="Select__Select___2Gi9N a11y__focus-border--default___60AXp"
          data-test-id="cf-ui-select"
          id="experiment"
          name="experiment"
          required=""
        >
          <option
            data-test-id="cf-ui-select-option"
            value="-1"
          >
            Select Optimizely experiment
          </option>
          <option
            data-test-id="cf-ui-select-option"
            value="15049730511"
          >
            landing-page-hero (running)
          </option>
        </select>
        <svg
          class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD Select__Select__icon___OBmvS"
          data-test-id="cf-ui-icon"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M5.29289 8.29289C5.68342 7.90237 6.31658 7.90237 6.70711 8.29289L12 13.5858L17.2929 8.29289C17.6834 7.90237 18.3166 7.90237 18.7071 8.29289C19.0976 8.68342 19.0976 9.31658 18.7071 9.70711L12.7071 15.7071C12.3166 16.0976 11.6834 16.0976 11.2929 15.7071L5.29289 9.70711C4.90237 9.31658 4.90237 8.68342 5.29289 8.29289Z"
          />
        </svg>
      </div>
    </div>
    
    <hr
      class="css-14380qe"
    />
    <h2
      class="Heading__Heading___83D3K"
      data-test-id="cf-ui-heading"
    >
      Variations:
    </h2>
    <div
      class="css-1pfvl5o"
    >
      <p
        class="Paragraph__Paragraph___2aO-9"
        data-test-id="cf-ui-paragraph"
      >
        Content created in this experiment is only available for this experiment.
      </p>
      <div
        class="css-i51og3"
      >
        <h2
          class="Subheading__Subheading___2mA9j css-1ih9qgr"
          data-test-id="cf-ui-subheading"
        >
          variation_1
           
          <small>
            (
            50
            % of traffic)
          </small>
        </h2>
        <div
          class="css-15g2oxy"
        >
          <div
            class="css-cet0rr"
          >
            <div
              class="Dropdown__Dropdown___nAsJ-"
              data-test-id="cf-ui-dropdown"
            >
              <button
                aria-expanded="false"
                aria-haspopup="menu"
                class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 TextLink__TextLink--primary___2Vc9F"
                data-test-id="cf-ui-text-link"
                type="button"
              >
                <span
                  class="TabFocusTrap__TabFocusTrap___39Vty"
                  tabindex="-1"
                >
                  <div
                    class="TextLink__TextLink__icon-wrapper___25_tI"
                  >
                    <svg
                      class="Icon__Icon___38Epv Icon__Icon--small___1yGZK undefined TextLink__TextLink__icon___3ggiB"
                      data-test-id="cf-ui-icon"
                      height="1em"
                      viewBox="0 0 24 24"
                      width="1em"
                    >
                      <path
                        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                      />
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                    </svg>
                  </div>
                  Create entry and link
                </span>
              </button>
            </div>
          </div>
          <div
            class="css-cet0rr"
          >
            <button
              class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 TextLink__TextLink--primary___2Vc9F"
              data-test-id="cf-ui-text-link"
              type="button"
            >
              <span
                class="TabFocusTrap__TabFocusTrap___39Vty"
                tabindex="-1"
              >
                <div
                  class="TextLink__TextLink__icon-wrapper___25_tI"
                >
                  <svg
                    class="Icon__Icon___38Epv Icon__Icon--small___1yGZK undefined TextLink__TextLink__icon___3ggiB"
                    data-test-id="cf-ui-icon"
                    height="1em"
                    viewBox="0 0 24 24"
                    width="1em"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z"
                    />
                  </svg>
                </div>
                Link an existing entry
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="css-i51og3"
      >
        <h2
          class="Subheading__Subheading___2mA9j css-1ih9qgr"
          data-test-id="cf-ui-subheading"
        >
          variation_2
           
          <small>
            (
            50
            % of traffic)
          </small>
        </h2>
        <div
          class="css-15g2oxy"
        >
          <div
            class="css-cet0rr"
          >
            <div
              class="Dropdown__Dropdown___nAsJ-"
              data-test-id="cf-ui-dropdown"
            >
              <button
                aria-expanded="false"
                aria-haspopup="menu"
                class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 TextLink__TextLink--primary___2Vc9F"
                data-test-id="cf-ui-text-link"
                type="button"
              >
                <span
                  class="TabFocusTrap__TabFocusTrap___39Vty"
                  tabindex="-1"
                >
                  <div
                    class="TextLink__TextLink__icon-wrapper___25_tI"
                  >
                    <svg
                      class="Icon__Icon___38Epv Icon__Icon--small___1yGZK undefined TextLink__TextLink__icon___3ggiB"
                      data-test-id="cf-ui-icon"
                      height="1em"
                      viewBox="0 0 24 24"
                      width="1em"
                    >
                      <path
                        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                      />
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                    </svg>
                  </div>
                  Create entry and link
                </span>
              </button>
            </div>
          </div>
          <div
            class="css-cet0rr"
          >
            <button
              class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 TextLink__TextLink--primary___2Vc9F"
              data-test-id="cf-ui-text-link"
              type="button"
            >
              <span
                class="TabFocusTrap__TabFocusTrap___39Vty"
                tabindex="-1"
              >
                <div
                  class="TextLink__TextLink__icon-wrapper___25_tI"
                >
                  <svg
                    class="Icon__Icon___38Epv Icon__Icon--small___1yGZK undefined TextLink__TextLink__icon___3ggiB"
                    data-test-id="cf-ui-icon"
                    height="1em"
                    viewBox="0 0 24 24"
                    width="1em"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z"
                    />
                  </svg>
                </div>
                Link an existing entry
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EditorPage should show the preemtive reconnect warning box 1`] = `
<a
  class="css-1lglquy"
  data-test-id="preemptive-connect"
>
  connect with Optimizely.
</a>
`;

exports[`EditorPage should show the reauth modal when no client is available 1`] = `
<p
  class="Paragraph__Paragraph___2aO-9 css-acwcvw"
  data-test-id="reconnect-optimizely"
>
  Your Optimizely session has expired. Reconnect to continue editing.
</p>
`;
